SSA/Core/MLIRSyntax/Parser.lean

```diff
@@ -0,0 +1,136 @@
+  | Except.error e => e.toMessageData.toString
+
+unsafe def elabIntoEIO {α : Type} : Lean.Environment → Lean.Expr → Lean.Syntax → EIO ParseError α :=
+  fun env typeName stx => do
+  let resE : EIO Exception α := elabIntoCore (α := α) typeName stx |>.run' {fileName := "parserHack", fileMap := default} {env := env}
+  -- TODO: I have no idea why the line below changes resE...
+  -- nor how to do this properly because `MessageData.toString` lives in `IO` for some reason.
+  --let errMsg : String ← resE.toIO' >>= printException |>.toEIO (fun e => s!"IO Error: {e}")
+  resE.adaptExcept (fun _ => s!"error elaborating")
+
+def region0Expr := (Expr.app (Expr.const `MLIR.AST.Region []) (Expr.const `Nat.zero []))
+def op0Expr := (Expr.app (Expr.const `MLIR.AST.Op []) (Expr.const `Nat.zero []))
+/--
+  Parse `Lean.Syntax` into `MLIR.AST.Region`.
+
+  This runs the Lean frontend stack and adds it to the TCB.
+  It should be safe (if we trust the lean parser) since we ensure that the types match: see
+  https://leanprover-community.github.io/mathlib4_docs/Std/Util/TermUnsafe.html#Std.TermUnsafe.termUnsafe_
+ -/
+def elabRegion (env : Lean.Environment) (stx : Lean.Syntax) : EIO ParseError (MLIR.AST.Region 0) := do
+  let reg ← unsafe elabIntoEIO (α := MLIR.AST.Region 0) env region0Expr stx
+  return reg
+
+/--
+  Parse `Lean.Syntax` into `MLIR.AST.Op`.
+
+  This runs the Lean frontend stack and adds it to the TCB.
+  It should be safe (if we trust the lean parser) since we ensure that the types match: see
+  https://leanprover-community.github.io/mathlib4_docs/Std/Util/TermUnsafe.html#Std.TermUnsafe.termUnsafe_
+ -/
+def elabOp (env : Lean.Environment) (stx : Lean.Syntax) : EIO ParseError (MLIR.AST.Op 0) := do
+  let reg ← unsafe elabIntoEIO (α := MLIR.AST.Op 0) env op0Expr stx
+  return reg
+
+def mkParseFun (syntaxcat : Name)
+    (ntparser : Syntax → EIO ParseError ParseOutput)
+    (s : String) (env : Environment) : EIO ParseError ParseOutput := do
+  let syn : Except String Syntax := Parser.runParserCategory env syntaxcat s
+  match syn with
+    | .error msg => throw msg
+    | .ok syn => ntparser syn
+
+private def mkNonTerminalParser (syntaxcat : Name)
+    (ntparser : Syntax → EIO ParseError ParseOutput)
+    (env : Environment) (s : String) : EIO ParseError ParseOutput :=
+  let parseFun := mkParseFun syntaxcat ntparser
+  parseFun s env
+
+-- TODO: do we need this copy of the environment, or can I remove it from one of the two?
+def regionParser : @ParseFun (MLIR.AST.Region 0) :=
+  fun env : Lean.Environment => mkNonTerminalParser `mlir_region (elabRegion env) env
+
+private def parseFile (env: Lean.Environment)
+    (parser: @ParseFun ParseOutput)
+    (filepath: System.FilePath) : IO (Option ParseOutput) := do
+  let lines ← IO.FS.lines filepath
+  let fileStr := lines.foldl (λ s₁ s₂ => s₁ ++ "\n" ++ s₂) "" -- TODO: use intercalate.
+  -- IO.println s!"[DEBUG] fileStr=\n{fileStr}"
```

Comment:
Leftover debug statement?

Reply:
yeah, I like to keep those around in case things break again :sweat_smile: don't know what the ideal way is to do that

Reply:
I guess I would either just leave it in the git history, add a comment why you want to leave this, or make this code that is disabled on a flag and set the flag to off (and add a comment there).

Reply:
OK, I removed it. Just please don't squash this PR when merging so that the git history is not lost.

Reply:
Do you not feel the history in this PR is sufficient? I feel this will make it significantly harder to actually see the real change in the history, will add commits-that-do-not-compile to the history, and also will randomly change our base std library back and forth. I am curious why the history in the PR is not sufficient for someone who wants to investigate this change?

Reply:
because if we squash this PR then we lose this history. It will be only on github and on some random stale branch that's not in the project's history. So if we're working on a file later and want to figure out what happened, why something was there or look at an older version of it, we cannot do 'git log path/to/file.lean`, we have to look it up in every single such PR branch, and that makes it completely unviable in my opinion. 

In my personal opinion `git log` and `git blame` etc are significantly more valuable than `git bisect`, and accurate, fine-grained helps more than a curated history where CI passes for every commit. I could see why for a very large project written in a chaotic language like C++ these priorities would be different, but for an academic project with a much lower PR/commit volume, in a strong typesystem, I think the value of the history is higher (I've never needed to bisect anything in this project, have you?)

Reply:
I feel that `git blame` is a lot more efficient as it points to the single PR that had a functional change to the repo rather than hitting some of these 42 commits and having to figure it out. In any case, this seems to need further discussion. Happy to have these eventually and not block this PR over our different perspectives here.